go/arrow/compute/exprs/exec.go

```diff
@@ -0,0 +1,631 @@
+		if err != nil {
+			return nil, err
+		}
+		return compute.NewDatum(scalar.MakeNullScalar(dt)), nil
+	case *expr.ListLiteral:
+		var elemType arrow.DataType
+
+		values := make([]scalar.Scalar, len(v.Value))
+		for i, val := range v.Value {
+			d, err := literalToDatum(mem, val, ext)
+			if err != nil {
+				return nil, err
+			}
+			defer d.Release()
+			values[i] = d.(*compute.ScalarDatum).Value
+			if elemType != nil {
+				if !arrow.TypeEqual(values[i].DataType(), elemType) {
+					return nil, fmt.Errorf("%w: %s has a value whose type doesn't match the other list values",
+						arrow.ErrInvalid, v)
+				}
+			} else {
+				elemType = values[i].DataType()
+			}
+		}
+
+		bldr := array.NewBuilder(memory.DefaultAllocator, elemType)
+		defer bldr.Release()
+		if err := scalar.AppendSlice(bldr, values); err != nil {
+			return nil, err
+		}
+		arr := bldr.NewArray()
+		defer arr.Release()
+		return compute.NewDatum(scalar.NewListScalar(arr)), nil
+	case *expr.MapLiteral:
+		dt, _, err := FromSubstraitType(v.Type, ext)
+		if err != nil {
+			return nil, err
+		}
+
+		mapType, ok := dt.(*arrow.MapType)
+		if !ok {
+			return nil, fmt.Errorf("%w: map literal with non-map type", arrow.ErrInvalid)
+		}
+
+		keys, values := make([]scalar.Scalar, len(v.Value)), make([]scalar.Scalar, len(v.Value))
+		for i, kv := range v.Value {
+			k, err := literalToDatum(mem, kv.Key, ext)
+			if err != nil {
+				return nil, err
+			}
+			defer k.Release()
+			scalarKey := k.(*compute.ScalarDatum).Value
+
+			v, err := literalToDatum(mem, kv.Value, ext)
+			if err != nil {
+				return nil, err
+			}
+			defer v.Release()
+			scalarValue := v.(*compute.ScalarDatum).Value
+
+			if !arrow.TypeEqual(mapType.KeyType(), scalarKey.DataType()) {
+				return nil, fmt.Errorf("%w: key type mismatch for %s, got key with type %s",
+					arrow.ErrInvalid, mapType, scalarKey.DataType())
+			}
+			if !arrow.TypeEqual(mapType.ValueType(), scalarValue.DataType()) {
+				return nil, fmt.Errorf("%w: value type mismatch for %s, got key with type %s",
+					arrow.ErrInvalid, mapType, scalarValue.DataType())
+			}
+
+			keys[i], values[i] = scalarKey, scalarValue
+		}
+
+		keyBldr, valBldr := array.NewBuilder(mem, mapType.KeyType()), array.NewBuilder(mem, mapType.ValueType())
+		defer keyBldr.Release()
+		defer valBldr.Release()
+
+		if err := scalar.AppendSlice(keyBldr, keys); err != nil {
+			return nil, err
+		}
+		if err := scalar.AppendSlice(valBldr, values); err != nil {
+			return nil, err
+		}
+
+		keyArr, valArr := keyBldr.NewArray(), valBldr.NewArray()
+		defer keyArr.Release()
+		defer valArr.Release()
+
+		kvArr, err := array.NewStructArray([]arrow.Array{keyArr, valArr}, []string{"key", "value"})
+		if err != nil {
+			return nil, err
+		}
+		defer kvArr.Release()
+
+		return compute.NewDatumWithoutOwning(scalar.NewMapScalar(kvArr)), nil
+	case *expr.StructLiteral:
+		fields := make([]scalar.Scalar, len(v.Value))
+		names := make([]string, len(v.Value))
```

Comment:
I didn't think the struct literal had names?  Or maybe you're making empty strings for names here?

Reply:
correct, it just uses empty strings for the names.